dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/JarDumper.java

```diff
@@ -0,0 +1,111 @@
+/*
+ * Copyright 2018 Google LLC.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.google.cloud.tools.opensource.classpath;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.reflect.ClassPath.ClassInfo;
+import java.io.IOException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+import org.apache.bcel.Const;
+import org.apache.bcel.classfile.Attribute;
+import org.apache.bcel.classfile.ConstantPool;
+import org.apache.bcel.classfile.InnerClass;
+import org.apache.bcel.classfile.InnerClasses;
+import org.apache.bcel.classfile.JavaClass;
+import org.apache.bcel.util.ClassPath;
+import org.apache.bcel.util.SyntheticRepository;
+
+/**
+ * This class reads a jar file and lists method references in its class files.
+ */
+class JarDumper {
+
+  /**
+   *  Lists all external method references from the jar file.
+   *
+   * @param jarFileUrl URL for the jar file
+   * @return list of the method signatures with their fully-qualified classes
+   * @throws IOException when there is a problem in reading the jar file
+   * @throws ClassNotFoundException when a class visible by Guava's reflect was unexpectedly
+   * not found by BCEL API
+   */
+  public static List<FullyQualifiedMethodSignature> listExternalMethodReferences(
+      URL jarFileUrl) throws IOException, ClassNotFoundException {
+    List<FullyQualifiedMethodSignature> methodReferences = new ArrayList<>();
+
+    SyntheticRepository repository = SyntheticRepository.getInstance(
+        new ClassPath(jarFileUrl.getFile()));
```

Comment:
Not sure this is right. getFile returns the path component of the URL. I think you want a java.io.File for the correct location on the file system. This may work in tests but if so it's very dependent on the current working directory.

Reply:
Let me think better type.

BCEL expects a file, while Guava needs a class loader, where I picked up URLClassLoader that needs URL.

Reply:
Now it's

```
  static List<FullyQualifiedMethodSignature> listExternalMethodReferences(
      File jarFile) throws IOException, ClassNotFoundException {
    List<FullyQualifiedMethodSignature> methodReferences = new ArrayList<>();
    Set<String> internalClassNames = new HashSet<>();

    String fileName = jarFile.getAbsolutePath();
    SyntheticRepository repository = SyntheticRepository.getInstance(new ClassPath(fileName));

    URL jarFileUrl = jarFile.toURI().toURL();
    for (ClassInfo classInfo : listTopLevelClassesFromJar(jarFileUrl)) {

```